### ISSUE TITLE
nohoist to ignore a package instead of ignore modules

### ISSUE THREAD

**Issue by huchenme** [yarnpkg/yarn#5833]
Hi,

I have a repo with a few sub packages, one of them (package1) got `react-scripts` 1.1.0 which use older version of Jest, babel and babel-jest, another sub package (package2) added dependency of newer version of Jest, babel and babel-jest. 

If I use 

```
"nohoist": [
   "**/react-scripts/",
   "**/react-scripts/**",
]
```

Jest, babel and babel-jest won't be hoisted at all to the root which make the packages using package 2 breaks.

If I don't put `nohoist` option, for some reason the hoist packages are the lower versions from `react-scripts`.

I am not sure what is the best option to resolve this, but ideally I just want to have package 1 nohoist (keep it's sub packages within its own node_modules)

**Comment by connectdotz** [comment]
try:
```
"nohoist": [
   "package1/react-scripts/",
   "package1/react-scripts/**",
]
```

**Comment by huchenme** [comment]
@connectdotz it does not work with that. As react-scripts got `babel-core`, `babel-jest` and `jest` dependencies. Those dependencies will not be hoisted to root if I apply above code.

What I want to achieve is that the dependencies in package 1 stay in its own node_modules, the other packages hoist as normal

**Comment by connectdotz** [comment]
if you want to disable hoist for **all** modules under package1, you can simply use 
```
nohoist": [
   "package1/**"
]
```

if you want to disable specific modules under package1, such as `react-native`, use the one I mentioned above.

nohoist just perform [pattern match](https://yarnpkg.com/blog/2018/02/15/nohoist/), there is no black magic, feel free to experiment...

**Comment by huchenme** [comment]
@connectdotz Thanks for the reply, but it is still not working.

By using

```
nohoist": [
   "package1/**"
]
```

my package2 will not hoist all dependencies that are under `package1/**`, without the `nohoist` option, they were all hoisted to root `node_modules`, but with `nohost` set to `package1/**`, anything included in `react-scripts` or deep dependencies are not hoist to root in package2

**My expectation**

```
nohoist": [
   "package1" (or anything we call it)
]
```

should not affect `package2`, it should just not hoist dependencies in `package1` but keep existing behavior of `package2` and other packages unchanged.

**Comment by connectdotz** [comment]
hmmm... make sure your `nohoist` config is in the root package.json and not in the workspace's. What you asked for is not unusual and it should work if configured correctly. If you still have any problem getting it to work, it is best to create a sample repo so we can help you further.

**Comment by huchenme** [comment]
What you are saying is true, but I am suggesting a feature that if we can disable `hoisting` for some packages without affecting other packages at all. Introducing `react-scripts` into `package1` have clearly breaks `package2` (I could manually spend a whole day to find all packages which break and manually add them but I don't think it should be that way).

Should we consider to change current nohoist behaviour to disable hoisting for certain packages, or add such feature to nohoist? Or can we just disable hoist for all packages but still use yarn as the package manager?

**Comment by connectdotz** [comment]
> Introducing react-scripts into package1 have clearly breaks package2
how?

**Comment by Pranit-Harekar** [comment]
^^ Same issue. I did this,
```json
 "nohoist": [
      "package-x**/",
      "**/package-x",
      "**/package-x/**"
    ]
```

It does not work as expected!!!! Dependencies still hoist to the root. @connectdotz  Any thoughts? Why it's happening? 😕

**Comment by tommedema** [comment]
@mtford90's question is already 5 months old without any feedback. Are people not experiencing issues from this? In my case it seems impossible to use yarn workspaces together with certain modules, especially react-scripts. Having trouble reproducing it in a small repo due to the complexities of the dependency trees with nohoist but it seems related to the above discussion.

**Comment by mtford90** [comment]
@tommedema :(

If anybody is interested, the "solution" I eventually ended up with was:

1) Ditch yarn + switch back to NPM.
2) Use lerna to manage the packages, hoisting etc.
3) Use [install-local](https://www.npmjs.com/package/install-local) for those packages that can't handle sym links e.g. react native with the metro bundler.

I look forward to the day that yarn workspaces has the flexibility needed to handle theses situations.

**Comment by arcanis** [comment]
I said it before, I'll say it again: **nohoist is a workaround**. If a package **needs** nohoist to work, then this package is broken and this should be reported to its maintainers. It will eventually break - it won't be ported over to the v2, for example.

In this particular case someone needs investigate how to fix the problems in RN that make it rely on hoisting that much. This is the real fix - nohoist isn't.

**Comment by huchenme** [comment]
@arcanis, I might have missed your message but I have not seen that message in this thread before. I spent so much time in this thread to create repo and explain why we need a feature, I my opinion, nohoist feature is incomplete at the moment that is why we are requesting a feature a make it easier to use.

If nohoist option is not right at the first place, I think we should remove this option in yarn, as you said, the package breaks and people should fix the problems themselves 

Additionally, I don’t think anything in my example is break and should fix, if you know how to fix it, please let me know

**Comment by arcanis** [comment]
There are various threads about nohoist. It's a confusing feature even for myself and I'm not even too sure to understand how it works - the hoisting part of the v1 codebase is quite messy and there's a lot of corner cases there. Part of why we've redesigned all this in the v2 😉

> If nohoist option is not right at the first place, I think we should remove this option in yarn, as you said, the package breaks and people should fix the problems themselves

It won't be part of the next major release (even more so since we won't generate node_modules at all). I don't think we'll remove it from the v1 though since some setups depend on it and removing it would be semver-major.

> Additionally, I don’t think anything in my example is break and should fix, if you know how to fix it, please let me know

What would be broken according to your post is that Jest cannot detect `jest-environment-jsdom` if it's being hoisted (is that still the case? I think it should work, btw 🤔). Whether a package is hoisted or not is an implementation detail and you shouldn't assume anything about it. In fact, you shouldn't even assume that there will be a `node_modules`. The only thing that will always be correct is acquiring dependencies through `require` and `require.resolve`.

**Comment by mtford90** [comment]
Agreed on it being removed if no intention to support these kind of issues. Anyone who tries to use this feature is probably found to hit this roadblock eventually.

A system that relies on the good behaviour (in this case, the good behaviour library maintainers) is probably destined for failure.

**Comment by connectdotz** [comment]
> A system that relies on the good behaviour (in this case, the good behaviour of library maintainers) is probably destined for failure.

ouch, that hurts, although I am guilty as charged... sorry for not being able to come back to this issue earlier. 

~10 months ago, Instead of adding more code to `nohoist`, I started to wonder if this is a self-inflicted wound due to hoisting, and if there is a better way to "hoist" so we won't break packages thus no need for "nohoist"... Discussed with @arcanis and he mentioned the v2 is on the way that should make this a moot point, I sort of write off this issue from my conscience since... 

ok, so a quick fix that makes "version-conflict-resolution works better with nohoist" is possible but I am hesitated to embark on such journey if v2 is on the way here soon or if the core team think we should not support "nohoist" in v1 code base anymore... thoughts?

**Comment by mtford90** [comment]
@connectdotz that wasn't aimed at you or yarn - but rather at the package maintainers who's packages are causing `nohoist` to break ;) - e.g. react & react-native. I meant that yarn/nohoist shouldn't depend on the good behaviour of packages that are installed using it.

Sorry for the perceived criticism and thanks for the explanation!

**Comment by arcanis** [comment]
> A system that relies on the good behaviour (in this case, the good behaviour of library maintainers) is probably destined for failure.

For the record I don't necessarily think of it as a matter of "good" versus "bad" behavior (as tempting as it might sometimes be). It's just that in the past years and until a few months ago package managers weren't enforcing good practices, and if you don't teach then you can't expect your users to learn.

Starting with PnP we've been much stricter than we previously were, and since PnP is the default install strategy from the v2 onwards it will become easier for library authors to spot problematic patterns. It's unfortunately a bit painful in some cases, but I hope the long-term benefit will outweight the cost.

Also note that those projects are just like Yarn in that they usually rely on their community's help to fix issues - so if you're familiar with for example react-native, I'm sure they would be glad if you could investigate what would be required for everything to work out of the box!
